api/envoy/config/core/v3/socket_option.proto

```diff
@@ -74,6 +82,18 @@ message SocketOption {
   // The state in which the option will be applied. When used in BindConfig
   // STATE_PREBIND is currently the only valid value.
   SocketState state = 6 [(validate.rules).enum = {defined_only: true}];
+
+  // Apply the socket option to the `stream socket type <https://linux.die.net/man/2/socket>`_.
+  // If both :ref:`stream <envoy_v3_api_field_config.core.v3.SocketOption.stream>` and
+  // :ref:`datagram <envoy_v3_api_field_config.core.v3.SocketOption.datagram>` fields are not
+  // specified, the socket option will be applied to all socket types.
+  Stream stream = 7;
+
+  // Apply the socket option to the `datagram socket type <https://linux.die.net/man/2/socket>`_.
+  // If both :ref:`stream <envoy_v3_api_field_config.core.v3.SocketOption.stream>` and
+  // :ref:`datagram <envoy_v3_api_field_config.core.v3.SocketOption.datagram>` fields are not
+  // specified, the socket option will be applied to all socket types.
+  Datagram datagram = 8;
```

Comment:
I think it is better to add both to a single message called SocketType:
```
message SocketType {
  message Stream {}
  message Datagram {}

  Stream stream = 1;
  Datagram datagram = 2;
}
```
And clarify what happens when none of the 2 exist, if one of the 2 exist, or if both exist (precedence).

Reply:
Done.